protocol/lib/eth/util.go

```diff
@@ -14,33 +16,41 @@ import (
 	ethcoretypes "github.com/ethereum/go-ethereum/core/types"
 )
 
-var bridgeEventAbi *ethabi.ABI
-
 const (
-	MIN_ADDRESS_BYTES = 20
-	MAX_ADDRESS_BYTES = 32
+	MinAddrLen = 20
+	MaxAddrLen = 32
 )
 
+// ABI Singleton and mutex.
+var bridgeEventAbi *ethabi.ABI
+var once sync.Once
+
 // getBridgeEventAbi returns the ABI (application binary interface) for the Bridge contract.
-func getBridgeEventAbi() *ethabi.ABI {
-	// Initialize the singleton if it does not exist.
-	if bridgeEventAbi == nil {
+func GetBridgeEventAbi() *ethabi.ABI {
+	// Initialize the singleton in a thread-safe way.
+	once.Do(func() {
 		bridgeAbi, err := ethabi.JSON(strings.NewReader(constants.BridgeEventABI))
 		if err != nil {
 			panic(err)
 		}
 		bridgeEventAbi = &bridgeAbi
+	})
+
+	// This could happen if the initialization fails in another thread.
+	if bridgeEventAbi == nil {
```

Comment:
Consider using [OnceValue](https://pkg.go.dev/sync#OnceValue) as it is simpler and guaranteed to panic when invoked instead of needing to check. This also allows you to combine the once with the value being stored and prevents erroneous usage of using `bridgeEventAbi` without the once.

Reply:
Unfortunately I don't think it exists in go 1.19, I think it's new as of 1.21

Reply:
We'll need to upgrade to 1.21 for Cosmos 0.50 so might as well and do it now: https://github.com/dydxprotocol/v4-chain/pull/586

Reply:
The upgrade to go 1.21 is now merged.